### ISSUE TITLE
Make it an error to have a leaf container layer once `buildScene` is called on a layer tree

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#90004]
Leaf container layers are permissible when building up the layer tree, but should not be present once `buildScene` is called. At best, this is some wasted cycles on some native calls to the SceneBuilder API. 

/cc @goderbauer @flar for any additional insights

**Comment by dnfield** [comment]
Updating to add: this should apply not only to leaf container layers, but subtrees containing only container layers.

**Comment by dnfield** [comment]
One possible exception may have to be `Leader/FollowerLayer`s, which `RenderEditable` uses to reserve spots in the tree. But even in this case, those layers should not be pushing engine layers in `addToScene`.

**Comment by dnfield** [comment]
I've done a few experiments and this does not seem to measurably help render times, including on lower end devices than what the devicelab has.

**Comment by flar** [comment]
Are you saying that we have no examples that trigger the condition? Or are you running a test case that intentionally creates a lot of empty leaf containers and it didn't seem to suffer any reduced render times?

**Comment by dnfield** [comment]
I'm saying it doesn't help on gallery transition perf.

**Comment by flar** [comment]
Have you gathered any metrics on how often this happens in the wild? Does Gallery have any cases of empty leaf containers?

**Comment by dnfield** [comment]
The places where it typically happens are RepaintBoundary render objects, which will push an OffsetLayer even if they have no children, and LeaderLayer objects, which are used as placeholders in the tree for other layers to find and paint into.

I've also seen at least one example of an OpacityLayer getting pushed with no children but an opacity, but that one is already special cased to not push to the engine if it has no children. We could probably just do that for all the other container layer types.

That said, what typically happens is that you have a few OffsetLayers with no children, but I haven't found a 'real' case of a large branch of the tree having no PictureLayer leaves.

**Comment by flar** [comment]
RepaintBoundary needs to do something to maintain its impact on isolating widgets from each other. By pushing the layer, it breaks the Picture boundaries as a side effect, but if it has no children, the same thing could be accomplished by simply calling stopRecordingIfNeeded I believe.

**Comment by dnfield** [comment]
Correct. I attempted something like `if the OffsetLayer has no children, remove it and stop recording`, but that did not seem to meaningfully move any benchmark - probably because it's only replacing a few quick operations with a different set of quick operations :)
